### DISCUSSION TITLE
Quando usar Flexbox e quando usar CSS Grid Layout.

### DISCUSSION THREAD

**Discussion by baltazarparra** [frontendbr/forum#105] [Dicas]
Queria compartilhar um material muito rico com vocês,
após assistir uma *palestra da [Rachel Andrew](https://rachelandrew.co.uk/) ,
fiquei com uma dúvida sobre quando usar Flexbox e quando usar Grid,
mas pouco tempo depois, ela mesma escreveu um artigo sobre isso.

https://rachelandrew.co.uk/archives/2016/03/30/should-i-use-grid-or-flexbox/

*[Palestra](https://www.youtube.com/watch?v=MXEzJ-IncX0l)

**Comment by lagden** [comment]
@baltazarparra da para ter o mesmo resultado usando `Grid` e ou `Flexbox`, Rachel não fez o pen direito!!

Original: http://codepen.io/rachelandrew/pen/YqqdXL
Fork: http://codepen.io/lagden/pen/aNygNN

O problema é que se usar `Grid` o seu cliente precisa ter o `Chrome Canary` como navegador!!
Sua pergunta é boa, para daqui uns anos!! :godmode:

**Comment by baltazarparra** [comment]
@lagden Concordo plenamente, mas a ideia era essa mesmo, levantar uma questão futura. A flexbox está bem a frente da Grid hoje, mas eu creio que em breve, estarão lado a lado.

Obs. o firefox dev também rola as grids.

**Comment by leandroruel** [comment]
por enquanto estou em um relacionamento com Flexbox! =D

**Comment by lagden** [comment]
Mas essa `draft` da `Grid Layout` é de 2011 e proposto pela `Microsoft` e não funciona no `Edge`!!

**Comment by tcelestino** [comment]
@lagden parece que não foi bem aceito pelo os demais.

Acho a proposta bem interessante, mas acredito que o flexbox precisa está mais presente, porque vejo que ainda tem muita gente que não conhece.

**Comment by lagden** [comment]
@tcelestino não conhece o `flexbox` ou o `grid`?

O `flexbox` é de 2009 e rapidamente foi implementado pela `Mozilla`, `Apple(webkit)` e `Opera(hoje também webkit)`, e hoje em dia nem precisa mais de prefixo para utilizá-lo!!

Por quê a `Microsoft` não fez o mesmo com o `grid`? :finnadie:

**Comment by felipefialho** [comment]
Compartilhando um artigo fresquinho.
- [Should I use Grid or Flexbox?](https://rachelandrew.co.uk/archives/2016/03/30/should-i-use-grid-or-flexbox/)

**Comment by tcelestino** [comment]
@lagden muita gente que trabalha com client-side não conhece o flexbox.

Sobre não se tornar popular, realmente é uma pena, acredito que faltou mais "empolgação" da própria MS.

**Comment by lagden** [comment]
@tcelestino não sei dizer se sua premissa é verdadeira!

A primeira vez que utilizei `flexbox` foi em 2010 e uma tela de `admin`. Só rolava bem Chrome, +- no Firefox e no IE, bahhh!! acho que tinham acabado de lançar o IE9...

Em 2013 foi quando coloquei em projeto para `enduser`, um pouco agressivo mas era um foda-se IE!!

**Bateu um flashback e agora acho que sua premissa seja verdadeira!!**

O pessoal de agência era ou é uma linha de montagem (copy/paste de código e muda o design). E claro, não podemos esquecer a era Macromedia Flash (quase morto!!) e o DOM(`Dreamweaver Ô Melhor`)!!!

Quando eu entrei na Tribo Interactive em 2012, os caras usavam reset do yahoo!!! e eu já no Sass!!!
E tinha que ficar escutando o @elvisDetona: `Eu não uso pre-processador, mimimimi...` uhahuahuha...

**Comment by elvisdetona** [comment]
[off]
@lagden  ahahahah, reset do yahoo.. "SEU AMIGO" que usava essa desgraça! E eu não gostava mesmo... mas eu não usava pra "hotsites e app do facebook", pq não precisava! 1 dia pra configurar o ambiente... nesse tempo eu já tinha terminado o trampo ahahaha.
E outra coisa... quando vc começar a pensar crossbrowser / semantica e acessibilidade, vc volta trocar idéia comigo ahahaha (e paga aquele burger no 12 tb heheheh)

**Comment by tarciozemel** [comment]
https://css-tricks.com/css-grid-replace-flexbox/

**Comment by felipefialho** [comment]
Ta fresquinho heim @diogomoretti ?

**Comment by diogomoretti** [comment]
Essa pergunta nós debatemos durante o evento de sábado após minha palestra. Tô pensando em fazer um vídeo para falar sobre e com exemplos.

**Comment by visantos** [comment]
Fala ai pessoal, atualmente estamos em junho de 2017 e vi que os navegadores, edge, chrome, firefox entre outros ganharam suporte para o grid layout, vocês acham que compensa pra galera que ta no inicio aprendendo a estruturar interfaces aprender o grid layout ao invés do flexbox? ou seria a mesma coisa aproveitar somente o sistema de grids do bootstrap, e rodar um estilo padrão nosso por cima?

**Comment by didiraja** [comment]
@Jovsantos voce parece ser iniciante como eu. Por isso, sempre acabo preferindo pegar as opções mais consolidadas, se essas novidades pegarem ou forem realmente úteis, daqui a um tempo ouviremos. Exatamente como foi/é com o flexbox.

**Comment by kvnol** [comment]
@Jovsantos @didiraja aprender é sempre bacana, mesmo que não que seja utilizado em produção pra valer, é legal já ter uma noção de como a parada toda funciona.

**Comment by didiraja** [comment]
@kvnol claro pra isso serve aqui o fórum, mas para coisas muito novas ou que não sei se precisarei mexer, eu acabo indo com cautela. Não evitar, apenas cautela, e ir lendo aos poucos artigos e relatos.

**Comment by kvnol** [comment]
@didiraja exato, concordo. Cautela para evitar dores de cabeça futuras! =D

**Comment by lagden** [comment]
https://github.com/frontendbr/forum/issues/105#issuecomment-210490154
@elvisdetona eu achei que você tinha mudado para Orlando/Flórida!!!
Bahh... só marcar que eu pago!!!

---

E falando sobre o assunto do flex e grid...
Grid é muito bom...

Apenas um teste...
https://codepen.io/lagden/pen/peKByR

Use os dois juntos e seja feliz!!!

**Comment by visantos** [comment]
@didiraja Realmente, sou iniciante, aprendi somente o básico sobre tags e propriedades css, mas por enquanto ainda não tenho noção de como criar algo de fato. Tenho muitas dúvidas em questão de estruturar a posição dos elementos no layout, então a solução que encontrei é usar somente o sistema de grids do bootstrap, vi que com o customize posso baixar somente a parte do grid, pelo menos até eu aprender a usar grid layout e/ou flexbox com eficiência.

PS: Sobre o Grid layout concordo com vc, é bom ter cautela, acho que daqui há 1 ano a galera ja vai estar usando ele com mais frequência, apesar das coisas mudarem sempre na web, acho que seria bom se essa questão de layout fosse padronizada.

**Comment by tgmarinho** [comment]
E ai passou um ano! opiniões?  rsrs

**Comment by lagden** [comment]
@tgmarinho agora tá tranquilo!

Todos os navegadores estão compatíveis...  
Eu utilizo o flexbox para algo simples e pequeno, pois o gridLayout é muito power!!!

**Por exemplo:**  
Esse form abaixo é 99% gridLayout, o único pedaço que usa flexbox é o botão de `Selecione os arquivos`

![exemplo](https://user-images.githubusercontent.com/130963/48521354-4c071280-e85b-11e8-9faa-7bbed375d04f.gif)

---

Esse post é bem bacana: https://hackernoon.com/the-ultimate-css-battle-grid-vs-flexbox-d40da0449faf

**Em suma:** Utilize os dois com sabedoria...

**Comment by LarissaAbreu** [comment]
Olá @tgmarinho , eu tenho utilizado os dois em conjunto na maioria das vezes...

Tem [esse artigo aqui](https://codepen.io/simoneas02/post/grid-layout) - da @simoneas02 que eu acho super bacana pq além de explicar a história e como o grid-layout funciona ele também fala rapidamente sobre em que momento utilizar cada uma das tecnologias :)

**Comment by Edvan27** [comment]
Bom, eu também sou um iniciante e fiz um curso bacana de flexbox que está disponível na net, e eu achei bem legal mesmo o flexbox, porém eu sei que preciso dar uma olhada no twitter bootstrap e essa parada de grid e tal, e para mobile por exemplo, eu curti bastante o materialize.

**Comment by leandroruel** [comment]
o criterio que uso para utilizar os 2:
CSS Grid >> Diagramação de Layout
Flexbox >> Alinhamento de Componentes de UI

**Comment by tgmarinho** [comment]
Legal, muito obrigado pelo feedback =) @LarissaAbreu @lagden 
Eu tinha visto https://caniuse.com/#search=grid%20layout ai acabei não usando, mas realmente o Grid Layout é mais tranquilo para algumas coisas. O que não consegui fazer com Flex Box foi pq eu não manjava mesmo, mas ai aprendi e consegui.

**Comment by faelplg** [comment]
De acordo com a própria especificação do CSS Grid sobre o Flexbox, "It is expected that both will be valuable and complementary tools for CSS authors." [CSS Grid Layout Module](https://drafts.csswg.org/css-grid/#intro). Logo, eles não foram desenvolvidos para se anularem, e sim para se complementarem. 

Depois de estudar bastante sistemas de layout, a melhor arquitetura que encontrei utilizando ambos na mesma estrutura foi a citada pelo @leandroruel , no qual o CSS Grid vai atuar melhor como layout da página, pois ele decide a distribuição bidimensional da mesma. O Flexbox, dentro de um layout bidimensional, vai se comportar ainda melhor, pois sua grande capacidade de se distribuir dentro de uma dimensão vai casar perfeitamente com as "caixas" desenhadas pelo grid.

**Comment by tarciozemel** [comment]
![flexbox-vs-cssgrid](https://user-images.githubusercontent.com/133777/49102826-fb32e900-f260-11e8-8b03-87b8556941ca.png)

**Comment by nanquim** [comment]
Esses vídeos são muito bons também:

[Flexbox vs. CSS Grid — Which is Better?](https://www.youtube.com/watch?v=hs3piaN4b5I&list=LLvgGzrRpwxT4cgWMlMa2ppA&index=19&t=0s)

[Using Flexbox + CSS Grid Together: Easy Gallery Layout](https://www.youtube.com/watch?v=dQHtT47eH0M&list=LLvgGzrRpwxT4cgWMlMa2ppA&index=17)
